oak_runtime/src/node/http/client.rs

```diff
@@ -0,0 +1,346 @@
+        }
+    }
+
+    /// Process an HTTP invocation for an external HTTP service.
+    async fn process_invocation(
+        &mut self,
+        runtime: &RuntimeProxy,
+        invocation: &Invocation,
+    ) -> Result<(), ProcessingError> {
+        // TODO(#1717): Update HTTP client metrics
+
+        // Receive a request from the invocation channel.
+        let request = invocation
+            .receive_request(&runtime)
+            .map_err(ProcessingError::ReadFailed)?;
+        debug!("Incoming HTTP request: {:?}", request);
+
+        let uri = request
+            .uri
+            .parse::<Uri>()
+            .map_err(|_err| ProcessingError::InvalidUri)?;
+
+        // Check that the client is allowed to handle the incoming request.
+        self.validate_uri(&uri)?;
+
+        let mut hyper_request_builder = hyper::Request::builder()
+            .method(&request.method[..])
+            .uri(uri);
+
+        if let Some(headers) = request.headers {
+            for (header_name, header_value) in headers.into_iter() {
+                hyper_request_builder = hyper_request_builder.header(header_name, header_value);
+            }
+        }
+        let hyper_request = hyper_request_builder
+            .body(hyper::Body::from(request.body))
+            .map_err(ProcessingError::BodyConversionError)?;
+
+        // Forward the request to the external HTTP service and wait for the response.
+        let resp = self
+            .http_client
+            .request(hyper_request)
+            .await
+            .map_err(ProcessingError::InterruptedWaitForResponse)?;
+
+        let mut response_handler = ResponseHandler::new(runtime.clone(), resp, invocation);
+        response_handler.handle().await
+    }
+
+    /// Check that the client is permitted to handle requests to the given URI.
+    fn validate_uri(&self, uri: &Uri) -> Result<(), ProcessingError> {
+        // Check that the scheme is present and is supported.
+        if uri.scheme().is_none() || *uri.scheme().unwrap() != self.scheme {
+            return Err(ProcessingError::UnsupportedScheme);
+        }
+        // Check that the `host` in the request URI is present, and matches the authority supported
+        // by the node, if the node's authority is non-empty.
+        if let Some(authority) = self.authority.clone() {
+            uri.authority()
+                .filter(|auth| *auth.to_owned() == authority)
+                .ok_or(ProcessingError::PermissionDenied)?;
+        }
+        Ok(())
+    }
+}
+
+struct ResponseHandler<'a> {
+    runtime: RuntimeProxy,
+    response: http::Response<hyper::Body>,
+    invocation: &'a Invocation,
+}
+
+impl<'a> ResponseHandler<'a> {
+    fn new(
+        runtime: RuntimeProxy,
+        response: http::Response<hyper::Body>,
+        invocation: &'a Invocation,
+    ) -> Self {
+        ResponseHandler {
+            runtime,
+            response,
+            invocation,
+        }
+    }
+
```

Comment:
I'm not sure I understand why ResponseHandler is a struct, it seems it may just be a top-level function, like `async fn handle_response(runtime, response, invocation)`? Anyways fine to keep as it is if you think it makes sense.